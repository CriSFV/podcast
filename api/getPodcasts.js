
export default async function getPodcasts() {
  const url = "https://itunes.apple.com/us/rss/toppodcasts/limit=100/genre=1310/json"
  const allorigins = encodeURIComponent(`${url}`);
  try{
    const response = await fetch(`https://api.allorigins.win/raw?url=${allorigins}`);
    const data = await response.json();
    return data;
  }
  catch(error) {
    throw new Error("It was an error:", error);
  };
}
