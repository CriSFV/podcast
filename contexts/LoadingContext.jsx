import { createContext, useState, useContext } from "react";

export const LoaderContext = createContext({});

export const useLoader = () => {
  const context = useContext(LoaderContext);
  if (!context) {
    throw new Error("useLoader must be used within a LoaderProvider");
  }
  return context;
};

export const LoaderProvider = ({ children }) => {
  const [loading, setLoading] = useState(false);

  const setLoadingState = (state) => {
    setLoading(state);
  };

  return (
    <LoaderContext.Provider value={{ loading, setLoadingState }}>
      {children}
    </LoaderContext.Provider>
  );
};
