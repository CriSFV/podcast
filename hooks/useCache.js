'use client'
import { useState } from "react";
import cacheService from "../api/cacheService";

export default function useCache(key, initialValue) {
  const [storedValue, setStoredValue] = useState(() => {
    try {
      const item = cacheService.get(key);
      return item ?? initialValue;
    } catch (error) {
      return initialValue;
    }
  });

  const setValue = (value) => {
    try {
      setStoredValue(value);
      cacheService.set(key, value);
    } catch (error) {
      console.log(error);
    }
  };

  const removeValue = (value) => {
    try {
      cacheService.remove(key);
    } catch (error) {
      console.log(error);
    }
  };
  return [storedValue, setValue, removeValue];
}
