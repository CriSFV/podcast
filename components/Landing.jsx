// import Layout from "../components/Layout";
import { useEffect, useState } from "react";
import getPodcasts from "../api/getPodcasts";
import store from "../api/store";
import Home from "./Home";
import Layout from "./Layout";
import { formatDataList } from "../helpers/formatDatafromApi";

function App() {
  const [data, setData] = useState([]);
  const [userSearch, setUserSearch] = useState("");
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    // check if more than one day has passed, if it's less than a day, get data from localStorage

    const date = store.get("date", "");
    if (Date.now() - date >= 86400000) {
      setIsLoading(true);
      getPodcasts().then((response) => {
        setIsLoading(false);
        const dataFormatted = formatDataList(response);
        setData(dataFormatted);
        store.clear();
        store.set("date", Date.now());
        store.set("podcastData", dataFormatted);
      });
    } else {
      setIsLoading(false);
      setData(store.get("podcastData", []));
    }
  }, []);

  // recive input value
  const handleSearch = (ev) => {
    setUserSearch(ev);
  };
  // receive podcast user selection
  const handleUserSelect = (ev) => {
    const podcast = data.find((pod) => pod.id === ev);
    podcast && store.set("podcastSelected", podcast);
  };

  const podcastFiltered =
    userSearch === ""
      ? data
      : data.filter(
          (podcast) =>
            podcast.title
              .toLocaleLowerCase()
              .includes(userSearch.toLocaleLowerCase()) ||
            podcast.author
              .toLocaleLowerCase()
              .includes(userSearch.toLocaleLowerCase())
        );

  return (
    <Layout isLoading={isLoading}>
      <div className="ppal_container">
        <Home
          data={podcastFiltered}
          handleSearch={handleSearch}
          handleUserSelect={handleUserSelect}
        />
      </div>
    </Layout>
  );
}

export default App;
