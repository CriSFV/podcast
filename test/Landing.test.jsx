import { render, screen, fireEvent } from "@testing-library/react";
import Landing from "../components/Landing";
import "@testing-library/jest-dom";
import { LoaderProvider } from "../contexts/LoadingContext";
import getPodcasts from "../api/getPodcasts";
import {
  podcastListMock,
  podcastListMockAfterFormated,
} from "./mocks/podcastsMock";
import useCache from "../hooks/useCache";

jest.mock("../hooks/useCache", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("../api/getPodcasts", () => jest.fn());
jest.mock("../api/getPodcastDetail", () => jest.fn());

jest.mock("next/router", () => ({
  ...jest.requireActual("next/router"),
  useRouter: () => ({
    query: {
      trackId: "1000604730376",
      id: "1535809342",
    },
  }),
}));

const LandingWithProvider = (props) => (
  <LoaderProvider>
    <Landing />
  </LoaderProvider>
);

describe("Landing", () => {
  beforeEach(() => {
    getPodcasts.mockClear();
    useCache.mockClear();
  });
  afterEach(() => {
    getPodcasts.mockRestore();
    useCache.mockRestore();
  });

  it("should match with snapshot", () => {
    useCache.mockImplementation(() => [Date.now() - 100]);
    useCache.mockImplementation(() => [podcastListMockAfterFormated]);

    const component = render(<LandingWithProvider />);
    expect(screen.getByText("Podcaster")).toBeInTheDocument();
    expect(component).toMatchSnapshot();
  });

  it("should render the list from localStorage", () => {
    const getPodcastMock = getPodcasts.mockImplementation(
      () => podcastListMock
    );
    useCache.mockImplementation(() => [Date.now() - 100]);
    useCache.mockImplementation(() => [
      podcastListMockAfterFormated,
      jest.fn(),
    ]);

    const component = render(<LandingWithProvider />);
    const podcastList = screen.getAllByRole("listitem");
    expect(screen.getByText("Podcaster")).toBeInTheDocument();
    expect(component.container).toHaveTextContent("Jhon Doe");
    expect(podcastList).toHaveLength(2);
    expect(getPodcastMock).not.toHaveBeenCalled();
  });

  it("should search a podcast from the input", () => {
    useCache.mockImplementation(() => [Date.now() - 100]);
    useCache.mockImplementation(() => [podcastListMockAfterFormated]);
    
    const userSearch = jest.fn();
    render(
      <LoaderProvider>
        <Landing userSearch={userSearch} />
      </LoaderProvider>
    );
    const input = screen.getByPlaceholderText("Filter podcast");
    const listCounter = screen.getByTestId("list-counter");
    expect(listCounter).toHaveTextContent(2);

    fireEvent.keyUp(input, { target: { value: "doe" } });

    expect(input.value).toBe("doe");
    expect(screen.getByText("Podcast Jhon Doe")).toBeInTheDocument();
    expect(screen.queryByText("The Mary Sister Podcast")).not.toBeInTheDocument();
    expect(listCounter).toHaveTextContent(1);
  });
});
