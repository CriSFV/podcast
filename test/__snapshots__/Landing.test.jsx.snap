// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ppal_container"
      >
        <header>
          <div
            class="loader__container"
          />
          <nav>
            <a
              class="title"
              href="/"
            >
              Podcaster
            </a>
          </nav>
          <hr
            class="separator"
          />
        </header>
        <main>
          <div
            class="ppal_container"
          >
            <section
              class="searcher"
            >
              <span
                class="searcher__results flex_column_center"
                data-testid="list-counter"
                id="list-counter"
              >
                2
              </span>
              <form
                action=""
              >
                <label
                  for="search"
                />
                <input
                  class="searcher__input"
                  id="search"
                  placeholder="Filter podcast"
                  type="text"
                />
              </form>
            </section>
            <section>
              <ul
                class="list"
              >
                <li
                  id="1535809341"
                >
                  <a
                    class="list__podcast"
                    href="/podcast/1535809341"
                  >
                    <img
                      alt="imagen_ The Mary Sister Podcast"
                      class="list__podcast__img"
                      data-nimg="1"
                      decoding="async"
                      height="120"
                      loading="lazy"
                      src="/_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=256&q=75"
                      srcset="/_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=256&q=75 2x"
                      style="color: transparent;"
                      width="120"
                    />
                    <div
                      class="list__podcast__text flex_column_space text__center"
                    >
                      <h6
                        class="list__podcast__text__title"
                      >
                        The Mary Sister Podcast
                      </h6>
                      <span
                        class="list__podcast__text__author"
                      >
                        Author: 
                        Mary Sister
                      </span>
                    </div>
                  </a>
                </li>
                <li
                  id="1535809342"
                >
                  <a
                    class="list__podcast"
                    href="/podcast/1535809342"
                  >
                    <img
                      alt="imagen_ Podcast Jhon Doe"
                      class="list__podcast__img"
                      data-nimg="1"
                      decoding="async"
                      height="120"
                      loading="lazy"
                      src="/_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=256&q=75"
                      srcset="/_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=256&q=75 2x"
                      style="color: transparent;"
                      width="120"
                    />
                    <div
                      class="list__podcast__text flex_column_space text__center"
                    >
                      <h6
                        class="list__podcast__text__title"
                      >
                        Podcast Jhon Doe
                      </h6>
                      <span
                        class="list__podcast__text__author"
                      >
                        Author: 
                        Jhon Doe
                      </span>
                    </div>
                  </a>
                </li>
              </ul>
            </section>
          </div>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ppal_container"
    >
      <header>
        <div
          class="loader__container"
        />
        <nav>
          <a
            class="title"
            href="/"
          >
            Podcaster
          </a>
        </nav>
        <hr
          class="separator"
        />
      </header>
      <main>
        <div
          class="ppal_container"
        >
          <section
            class="searcher"
          >
            <span
              class="searcher__results flex_column_center"
              data-testid="list-counter"
              id="list-counter"
            >
              2
            </span>
            <form
              action=""
            >
              <label
                for="search"
              />
              <input
                class="searcher__input"
                id="search"
                placeholder="Filter podcast"
                type="text"
              />
            </form>
          </section>
          <section>
            <ul
              class="list"
            >
              <li
                id="1535809341"
              >
                <a
                  class="list__podcast"
                  href="/podcast/1535809341"
                >
                  <img
                    alt="imagen_ The Mary Sister Podcast"
                    class="list__podcast__img"
                    data-nimg="1"
                    decoding="async"
                    height="120"
                    loading="lazy"
                    src="/_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=256&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=256&q=75 2x"
                    style="color: transparent;"
                    width="120"
                  />
                  <div
                    class="list__podcast__text flex_column_space text__center"
                  >
                    <h6
                      class="list__podcast__text__title"
                    >
                      The Mary Sister Podcast
                    </h6>
                    <span
                      class="list__podcast__text__author"
                    >
                      Author: 
                      Mary Sister
                    </span>
                  </div>
                </a>
              </li>
              <li
                id="1535809342"
              >
                <a
                  class="list__podcast"
                  href="/podcast/1535809342"
                >
                  <img
                    alt="imagen_ Podcast Jhon Doe"
                    class="list__podcast__img"
                    data-nimg="1"
                    decoding="async"
                    height="120"
                    loading="lazy"
                    src="/_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=256&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fis1-ssl.mzstatic.com%2Fimage%2Fthumb%2FPodcasts113%2Fv4%2Ff2%2F21%2Ffa%2Ff221fabd-017f-5125-633b-f1fe4f39802a%2Fmza_182995249085044287.jpg%2F170x170bb.png&w=256&q=75 2x"
                    style="color: transparent;"
                    width="120"
                  />
                  <div
                    class="list__podcast__text flex_column_space text__center"
                  >
                    <h6
                      class="list__podcast__text__title"
                    >
                      Podcast Jhon Doe
                    </h6>
                    <span
                      class="list__podcast__text__author"
                    >
                      Author: 
                      Jhon Doe
                    </span>
                  </div>
                </a>
              </li>
            </ul>
          </section>
        </div>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
